deckhand/engine/secret_substitution.py

```diff
@@ -0,0 +1,126 @@
+                    name: some-certificate-key-asdf-1234
+            data:
+              tls_endpoint:
+                  certificate: null  # Data to be substituted.
+                  certificateKey: null  # Data to be substituted.
+        """
+        # Validate that data section exists.
+        try:
+            self.data['data']
+        except (KeyError, TypeError) as e:
+            raise errors.InvalidFormat(
+                'The provided YAML file has no data section: %s' % e)
+        # Validate that substitutions section exists.
+        try:
+            substitutions = self.data['metadata']['substitutions']
+        except (KeyError, TypeError) as e:
+            raise errors.InvalidFormat(
+                'The provided YAML file has no metadata/substitutions '
+                'section: %s' % e)
+
+        # Validate that "src" and "dest" fields exist per substitution entry.
+        error_message = ('The provided YAML file is missing the "%s" field '
+                         'for the %s substition.')
+        for s in substitutions:
+            if 'src' not in s:
+                raise errors.InvalidFormat(error_message % ('src', s))
+            elif 'dest' not in s:
+                raise errors.InvalidFormat(error_message % ('dest', s))
+
+        # Validate that each "dest" field exists in the YAML data.
+        destinations = [s['dest'] for s in substitutions]
+        sub_data = self.data['data']
+        for dest in destinations:
+            result, missing_attr = self._multi_getattr(dest, sub_data)
+            if not result:
+                raise errors.InvalidFormat(
+                    'The attribute "%s" included in the "dest" field "%s" is '
+                    'missing from the YAML data: "%s".' % (
+                        missing_attr, dest, sub_data))
+
+    def _multi_getattr(self, multi_key, substitutable_data):
+        """Iteratively check for nested attributes in the YAML data.
+
+        Check for nested attributes included in "dest" attributes in the data
+        section of the YAML file. For example, a "dest" attribute of
+        ".foo.bar.baz" should mean that the YAML data adheres to:
+
+        .. code-block:: yaml
+
+           ---
+           foo:
+               bar:
+                   baz: <data_to_be_substituted_here>
+
+        :param multi_key: A multi-part key that references nested data in the
+            substitutable part of the YAML data, e.g. ".foo.bar.baz".
+        :param substitutable_data: The section of data in the YAML data that
+            is intended to be substituted with secrets.
+        :returns: Tuple where first value is a boolean indicating that the
+            nested attribute was found and the second value is the attribute
+            that was not found, if applicable.
+        """
+        attrs = multi_key.split('.')
+        # Ignore the first attribute if it is "." as that is a self-reference.
+        if attrs[0] == '':
```

Comment:
Thanks, I'll look into which is more intuitive and make the updates in a follow-up patch.